### PR TITLE
dft: adding initial schema for DFT data sharing

### PR DESCRIPTION
**Pull request by fgaray**
This is a WIP but good to merge now.

If this looks good @maliberty , I will followup with the custom implementation of the getters/setters, but I want to first check that I am doing the right thing with the schema/json files.

Following your comment on: https://github.com/The-OpenROAD-Project/OpenROAD/discussions/4269#discussioncomment-7581928

Thanks!

### CODE REVIEW

**Comment by fgaray** [comment]
Also @maliberty Are we fixing the clang-tidy messages for auto-generated code?

**Comment by maliberty** [comment]
> Also @maliberty Are we fixing the clang-tidy messages for auto-generated code?

Don't worry about those.

**Comment by maliberty** [comment]
Something odd has happened here - you seem to have merged master but the diffs now show up as changes in this PR.  Did you use git merge?

**Comment by fgaray** [comment]
> Something odd has happened here - you seem to have merged master but the diffs now show up as changes in this PR. Did you use git merge?

I am not sure what I did. For some reason, I can't push my changes to this branch but I still have them on local. I may need to do a force push to sync this. Are you ok with me doing that? I know that it may cause some issues with the review here in github.

**Comment by maliberty** [comment]
Ideally I would avoid force pushing after a review has started.  However I think you should in this case things are messed up.

**Comment by maliberty** [comment]
Note there was a schema update yesterday so you'll need to resolve any conflicts.

**Comment by fgaray** [comment]
Hi @maliberty , thanks letting me force push this time. I think this is now ready for review. I am able to see the changes now.

File: src/odb/src/codeGenerator/schema/scan/dbScanPin.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    {
+      "name": "bterm",
+      "type": "dbId<dbBTerm>",
+      "flags": ["private"]
+    },
+    {
+      "name": "iterm",
+      "type": "dbId<dbITerm>",
+      "flags": ["private"]
+    }
```
Comment (maliberty) [inline src/odb/src/codeGenerator/schema/scan/dbScanPin.json:14]: Should this be a std::variant ?

File: src/odb/include/odb/db.h [hunk 2]
```diff
@@ -7524,6 +7528,38 @@ class dbPowerSwitch : public dbObject
...
+
+class dbScanPartition : public dbObject
+{
+ public:
+};
+
+class dbScanPin : public dbObject
+{
+ public:
+};
```
Comment (maliberty) [inline src/odb/include/odb/db.h:7561]: There is no way to create these objects or get/set any fields.  dbScanInst & dbScanChain  are only slightly better.  I think you need to add some custom methods here.

File: src/odb/src/codeGenerator/schema/scan/dbScanPin.json (hunk 1 above)
Comment (maliberty) [reply src/odb/src/codeGenerator/schema/scan/dbScanPin.json:14]: Yes

File: src/odb/include/odb/db.h (hunk 2 above)
Comment (maliberty) [reply src/odb/include/odb/db.h:7561]: Ok so long as you feel you won't need to change the schema again

File: src/odb/src/db/dbScanPin.cpp [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+#include "db.h"
+#include "dbDatabase.h"
+#include "dbDiff.hpp"
+#include "dbTable.h"
+#include "dbTable.hpp"
+
+// User Code Begin Includes
+namespace {
+template <class>
+inline constexpr bool always_false_v = false;
```
[the remark this answers is not in the document]
Comment (fgaray) [reply src/odb/src/db/dbScanPin.cpp:45]: I think this one is a false positive. If I remove this line I will get a compile error in line 168 in this file.

**Comment by fgaray** [comment]
Hi @maliberty , this is ready for another look.

I changed and implemented some methods in dbScanPin using std::variant, the rest of the schema is still empty as I plan to do them in a next PR. Could you please see my implementations for the serialization (>>, <<)?

Thanks!

**Review by maliberty — APPROVED** [review]

**Pull request merged**